compiler/semantic/expr.go

```diff
@@ -601,6 +609,33 @@ func (a *analyzer) semCall(call *ast.Call) dag.Expr {
 	}
 }
 
+func (a *analyzer) semCallDatePart(partExpr, argExpr ast.Expr) dag.Expr {
+	var part string
+	switch p := partExpr.(type) {
+	case *ast.ID:
+		part = p.Name
+	case *ast.Primitive:
+		if p.Type != "string" {
+			a.error(partExpr, fmt.Errorf("part must be an identifier or string"))
+			return badExpr()
+		} else {
+			part = p.Text
+		}
+	default:
+		a.error(partExpr, fmt.Errorf("part must be an identifier or string"))
```

Comment:
Why can't this be an arbitrary expression?

Reply:
Checked how postgres handles this. For `date_part` postgres allows this an arbitrary expression for the part argument. I'll change things to allow this and the dag for CallDatePart goes away. For the `extract` form you need to specify either an identifier or string and cannot have it be an arbitrary expression. If we want to get super granular here it throws a syntax error if you use any other expression so perhaps I should write up the peg to do this as well.

Reply:
I wouldn't bother making the grammar more complicated for the sake of producing the same error as PostgreSQL. We shouldn't accept a smaller language than it does but right now there's no reason we shouldn't accept a larger one.

Reply:
This will be a lot more code make this a generic expression because you can't assume the part expression is a const.